Implement OSD encryption key rotation

Rakshith-R opened the pull request:
<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/latest/Contributing/development-flow/)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**
This pr implements the following:

- [core: add KeyRotationSpec to cephcluster CR](https://github.com/rook/rook/commit/5dc8f5a688c72fb95dc032d001b8b7ed47e78cae) 

- [osd: add util functions to support encryption key rotation](https://github.com/rook/rook/commit/e3d9a3a13a507842ea252d1c9a7af40fb287c1e9) 

- [ceph: add support for updating secret for default KMS](https://github.com/rook/rook/commit/9f1ea407fb4a384b63e89bcb747f7d19a89428d1) 


- [osd: add rotate-key functionality to rook's key-management cmd](https://github.com/rook/rook/commit/8abba1bbd362ded53ab16446738b3c662e1efef7) 

- [ceph: add util function to create key rotation cron job](https://github.com/rook/rook/commit/20e7b5e453822f6bafb770612e08e763392ec1ec) 

- [ceph: add capability to reconcile key rotation cron jobs](https://github.com/rook/rook/commit/a9659b16e502d32d478faf85de2e8e5ddeeab967)

---

### Cephcluster CR changes

 parameters `enabled: <true|false>` and `schedule: <cron_format, default to @weekly>` are added to
`security.keyRotation` section of cephcluster spec here https://github.com/rook/rook/blob/master/Documentation/Storage-Configuration/Advanced/key-management-system.md.
```
security:
  kms:
    keyRotation: 
      enabled: "true"
      schedule: "@weekly" 
    connectionDetails:
      KMS_PROVIDER: vault
      ....
      tokenSecretName: rook-vault-token
```

---

### KMS changes

Support for `KMS.UpdateSecret()` is added.
Currently only default type/ k8s secret is supported
It will be extended in future to support other KMS such as vault,ibm hpcs, kmip etc.

---

### KEK Rotation logic would be as follows: 

| Step | Operation                 | Luks Slot 0 | Luks Slot 1 | Key in KMS |
|:---- |:------------------------- |:----------- |:----------- |:---------- |
| 1    | Obtain K1                 | K1          |             | K1         |
| 2    | Add K1 to slot 1          | K1          | K1          | K1         |
| 3    | Create K2 & add to slot 0 | K2          | K1          | K1         |
| 4    | Update K2 in KMS          | K2          | K1          | K2         |
| 5    | Remove K1 from slot 1     | K2          |             | K2         |
 
Note: The above steps will ensure the KEK in kms will be able to open the encrypted device even if the operation is disrupted at any step and all the edge cases from disrupted processes are handled.

`luksAddKey, luksChangeKey, luksKillSlot` commands are being used to achieve this.

Refer: [10 Linux cryptsetup Examples for LUKS Key Management (How to Add, Remove, Change, Reset LUKS encryption Key)](https://www.thegeekstuff.com/2016/03/cryptsetup-lukskey/)

---

### KEK Rotation Cron Job

- One Cron Job per encrypted PVC backed OSD is be created when key rotation is enabled with the given schedule.
- The Cron Job will use OSD pod affinity `requiredDuringScheduling..` using OSD's labels as selector to run on the same node as the OSD.
- The Cron Job will share the host bridge directory with the OSD which contains the enrcypted devices mapped to be able to rotate the KEK.

--- 


**Which issue is resolved by this Pull Request:**
Resolves #7925 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

Rakshith-R commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -268,6 +272,18 @@ type KeyManagementServiceSpec struct {
+	// Schedule represents the cron schedule for key rotation.
+	// +optional
+	// +kubebuilder:default="@weekly"
+	Schedule string `json:"schedule,omitempty"`
```
> How about an enum? Or is the string to leave it open to the CronJob schedule spec? Perhaps is there a CronJob schedule type?

It is a cron expression which cannot be validated by kubebuilder, 
yes we can leave it to the cronjob spec for validation when we create it.
https://en.wikipedia.org/wiki/Cron

Rakshith-R commented on pkg/daemon/ceph/osd/encryption.go:
```diff
@@ -173,6 +174,125 @@ func openEncryptedDevice(context *clusterd.Context, disk, target, passphrase str
+		disk,
+		slot,
+	}
+	logger.Debug("executing: ", cryptsetupBinary, strings.Join(args, " "))
```
> Could you show example logs from testing? Some logs like this don't seem formatted.

the executor actually logs these information, removing this line.

Rakshith-R commented on pkg/daemon/ceph/osd/encryption.go:
```diff
@@ -173,6 +174,125 @@ func openEncryptedDevice(context *clusterd.Context, disk, target, passphrase str
+	output, err := context.Executor.ExecuteCommandWithTimeout(luksOpenCmdTimeOut,
+		cryptsetupBinary, args...)
+	// ignore the error if the key slot is not active.
+	if err != nil && !strings.Contains(err.Error()+output, fmt.Sprintf("Keyslot %s is not active", slot)) {
```
> String compares in error messages can be brittle, is there any more specific API call we can query specifically if the slot is not active?

unfortunately no, we have to make do using `luksAddKey, luksKillSlot and luksChangeKey` cmds. 
https://man7.org/linux/man-pages/man8/cryptsetup.8.html

But The codepaths only rely on error strings to navigate edge cases and for any step to be complete 
the final luks command must return success. 

The edge cases are those where a key rotation operation stopped in between.
I've added extensive mock unit tests for alll these cases.

Rakshith-R commented on pkg/daemon/ceph/osd/kms/k8s.go:
```diff
@@ -45,6 +45,12 @@ func (c *Config) storeSecretInKubernetes(pvcName, key string) error {
 	}
 
+	// Set the ownerref to the Secret
+	err = c.ClusterInfo.OwnerInfo.SetControllerReference(s)
```
> Why did this move here instead of being inside `generateOSDEncryptedKeySecret()`?

moving it back. 
this was from my earlier attempts.

Rakshith-R commented on pkg/daemon/ceph/osd/kms/k8s.go:
```diff
@@ -54,6 +60,24 @@ func (c *Config) storeSecretInKubernetes(pvcName, key string) error {
+// updateSecretInKubernetes updates the dmcrypt key in a Kubernetes Secret
+func (c *Config) updateSecretInKubernetes(pvcName, key string) error {
+	secretName := GenerateOSDEncryptionSecretName(pvcName)
+	secret, err := c.context.Clientset.CoreV1().Secrets(c.ClusterInfo.Namespace).Get(c.ClusterInfo.Context, secretName, metav1.GetOptions{})
```
> Instead of a Get and Update, could we call `generateOSDEncryptedKeySecret()` to create the spec and directly call Update?

This call goes from the key rotation job.

It's better to update the already existing secret(with all the ownerref and the sought set already).

Rakshith-R replied:
> Do we ever expect anything to update the secret except rook? If rook completely owns the resources, it's a nice pattern to generate the same spec whether it's a create or update, but this way certainly works too.

This update runs from the cronjob, creating same spec also means setting ownerref to the cluster cr(needs extra rbac to do so). 
Just fetching and updating only the key section of the secret is simpler.

Rakshith-R commented on pkg/operator/ceph/cluster/osd/key_rotation.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// getKeyRotationPodTemplateSpec returns the pod template spec for the key rotation job.
+func (c *Cluster) getKeyRotationPodTemplateSpec(osdProps osdProperties, osd OSDInfo, restart v1.RestartPolicy) (*v1.PodTemplateSpec, error) {
```
> Some design questions:
> 
> * One cron job per OSD?

yes

> * Could we instead have a single cron job per node that rotates all OSDs on that node? This would just require that OSDs not be portable unless the job could dynamically discover which OSDs on the node need to be rotated.
> * Since we're using host paths, it will be critical to run the cron jobs on the same node where the OSDs are running.

The cronjob per OSD will have podAffinity set with requiredDuringSchedule using that OSD's labels as selector, therefore the corresponding cronjob pod will always be sheduled on the node where its OSD is running.
Each CronJob will have access to only that OSD's underlying bridge host path where `block*-tmp` encrypted devices are mapped onto and can run only if that folder exists.

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -71,13 +88,15 @@ func startSecret() *kms.Config {
-	err = kms.ValidateConnectionDetails(ctx, context, &cephCluster.Spec.Security.KeyManagementService, namespace)
-	if err != nil {
-		rook.TerminateFatal(errors.Wrap(err, "failed to validate kms connection details"))
+	if cephCluster.Spec.Security.KeyManagementService.IsEnabled() {
```
> Will we ever be in this code path if the KMS is not enabled? Or is this code path triggered in a different way than previously so we need to check for this now?

it'll be trigger when we call rotate secret, which currently only supports default type which isn't considered keymanagementservice and
does not need any connection details.

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -113,3 +132,91 @@ func getSecret(cmd *cobra.Command, args []string) {
+func cliRotateSecret() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "rotate-key [kms-secret-key] [data-device] [metadata-device] [wal-device]",
+		Short: "Rotate a secret from a given KMS",
```
> I thought we weren't rotating from a KMS yet?

modified wording,

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -113,3 +132,91 @@ func getSecret(cmd *cobra.Command, args []string) {
+func rotateSecret(cmd *cobra.Command, args []string) {
+	rook.SetLogLevel()
+	rook.LogStartupInfo(KeyManagementCmd.Flags())
+	// Initialize the context
```
:+1: done,

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -113,3 +132,91 @@ func getSecret(cmd *cobra.Command, args []string) {
+
+	// Ensure currentKey is in slot 1.
+	for _, devicePath := range devicePaths {
+		logger.Debugf("adding the current key to slot %q of the device %q", slotOne, devicePath)
```
:+1:

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -113,3 +132,91 @@ func getSecret(cmd *cobra.Command, args []string) {
+
+	logger.Debugf("generating new key")
+	// Generate new key.
+	newKey, err := oposd.GenerateDmCryptKey()
```
> 

yes, 
We always add the key from the KMS/ k8s secret to slot 1 in step 2.
So we can change slot 0 and still be able to open device using slot 1 seamlessly.

Rakshith-R commented on Documentation/Storage-Configuration/Advanced/key-management-system.md:
```diff
@@ -12,6 +13,12 @@ The `security` section contains settings related to encryption of the cluster.
     * `connectionDetails`: the list of parameters representing kms connection details
     * `tokenSecretName`: the name of the Kubernetes Secret containing the kms authentication token
+  * `keyRotation`: Key Rotation settings
+    * `enabled`: whether key rotation is enabled or not, default is `false`
```
> Is there a reason someone wouldn't want to rotate the keys or is there a risk? I wonder if we should default to true.

Currently only default type/k8s secret is supported and
I'd prefer we keep it false for a few cycles so there more tests/
fixes if any is needed.

Rakshith-R commented on Documentation/Storage-Configuration/Advanced/key-management-system.md:
```diff
@@ -12,6 +13,12 @@ The `security` section contains settings related to encryption of the cluster.
+    * `schedule`: the cron schedule for key rotation, default is `@weekly`
+
+!!! note
+    Currently key rotation is only supported for default type, where the Key Encryption Keys are stored in a Kubernetes Secret.
```
> So if they're using a KMS, the keys won't be rotated?

The job will fail with error KMS does not support secret update.

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -113,3 +132,91 @@ func getSecret(cmd *cobra.Command, args []string) {
+	// Ensure currentKey is in slot 1.
+	for _, devicePath := range devicePaths {
+		logger.Debugf("adding the current key to slot %q of the device %q", slotOne, devicePath)
+		err = osd.AddEncryptionKey(context, devicePath, currentKey, currentKey, slotOne)
```
> If the key is already in slot one, will this succeed?

yes, 
if key matches with newPassphrase, ensureEncryptionKey will return true indicating match
if there's no match, we kill the slot and add the newpassphrase there.

Rakshith-R commented on pkg/operator/ceph/cluster/osd/key_rotation.go:
```diff
@@ -0,0 +1,279 @@
+
+	// Replace default unreachable node toleration if the osd pod is portable and based in PVC
+	if osdProps.portable {
+		k8sutil.AddUnreachableNodeToleration(&podTemplateSpec.Spec)
```
> We shouldn't need this toleration for the cron job, just for daemon pods

I think it should apply to key rotation pod since it applies to OSD as well.
we want this pod to be eviction duration to be similar to that of OSD pod's.

Rakshith-R replied:
removed it :+1:

Rakshith-R commented:
Thanks for the quick and detailed review ! :smile: 

> Some overall needs:
> 
> * Unit tests

added some.

> * Integration tests

I'l try to add this soon,
if not possible in this, I'll do it in a follow-up pr.

> * Design documentation. Where can we see an overview of the key rotation? At least we need a final design in #7925 even if not a full design doc.

I've added the final design in a comment on the issue and in this pr description.

--- 

I've to still do the following things, I'll get to it on monday,
- changes to helm values.yaml, helm docs, cephcluster example yamls (any pointers on which of these yamls need key rotation examples ? )
- I'll go over again and see if I  can add unit tests in any other places too.
- intergration test if possible.

Please review the changes so far and let me know if anything else needs to modified or explained.

subhamkrai commented on pkg/daemon/ceph/osd/key_rotation.go:
```diff
@@ -0,0 +1,95 @@
+	slotOne  string = "1"
+)
+
+func RotateEncryptionKey(kms *kms.Config, context *clusterd.Context, secretName string, devicePaths []string) error {
```
```suggestion
func RotateEncryptionKey(context *clusterd.Context, kms *kms.Config, secretName string, devicePaths []string) error {
```
nit

subhamkrai commented on pkg/daemon/ceph/osd/key_rotation.go:
```diff
@@ -0,0 +1,95 @@
+	// Fetch the currentKey.
+	currentKey, err := kms.GetSecret(secretName)
+	if err != nil {
+		rook.TerminateFatal(errors.Wrapf(err, "failed to get secret %q", secretName))
```
question: is this right to use `TerminateFatal`? since it will do exit 1.

Rakshith-R replied:
missed it while moving it to osd package, 
changed to return

Rakshith-R commented:
> Thanks for the quick and detailed review ! smile
> 
> > Some overall needs:
> > 
> > * Unit tests
> 
> added some.
> 
> > * Integration tests
> 
> I'l try to add this soon, if not possible in this, I'll do it in a follow-up pr.
> 
> > * Design documentation. Where can we see an overview of the key rotation? At least we need a final design in [Implement encryption key rotation for cluster-wide (OSDs) encryption #7925](https://github.com/rook/rook/issues/7925) even if not a full design doc.
> 
> I've added the final design in a comment on the issue and in this pr description.
> 
> I've to still do the following things, I'll get to it on monday,
> 
> * changes to helm values.yaml, helm docs, cephcluster example yamls (any pointers on which of these yamls need key rotation examples ? )

added key rotation example in cluster-on-pvc.yaml (only this contained `.spec.security` section).

> * I'll go over again and see if I  can add unit tests in any other places too.

done.

> * intergration test if possible.

I've added this alongside canary-integration-test :   encryption-pvc-db-wal :. 
This will check for new key in secret and the next step of osd deployment removal will check that new key can open
the encrypted device. 

> * Add Logs from key rotation cron job run for reference.
> 
> Please review the changes so far and let me know if anything else needs to modified or explained.

<details>
<summary>

### click here to expand key rotation pod logs:
</summary>

```
runner@fv-az275-887:~/work/rook/rook$ k logs rook-ceph-osd-key-rotation-0-27958250-gmtfq -n rook-ceph
2023-02-27 10:50:12.292968 I | cephosd: fetching the current key
2023-02-27 10:50:12.296243 I | cephosd: adding the current key to slot "1" of the device "/var/lib/ceph/osd/block-tmp"
2023-02-27 10:50:12.296759 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/325846276 --key-slot=1 luksAddKey /var/lib/ceph/osd/block-tmp /tmp/1585851795
2023-02-27 10:50:36.971463 I | cephosd: adding the current key to slot "1" of the device "/var/lib/ceph/osd/block.db-tmp"
2023-02-27 10:50:36.971735 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/739593765 --key-slot=1 luksAddKey /var/lib/ceph/osd/block.db-tmp /tmp/1860780317
2023-02-27 10:50:46.211929 I | cephosd: adding the current key to slot "1" of the device "/var/lib/ceph/osd/block.wal-tmp"
2023-02-27 10:50:46.212323 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/2048944663 --key-slot=1 luksAddKey /var/lib/ceph/osd/block.wal-tmp /tmp/2041810508
2023-02-27 10:51:19.710455 I | cephosd: generating new key
2023-02-27 10:51:19.710497 I | cephosd: removing key slot "0", if found, of the device "/var/lib/ceph/osd/block-tmp"
2023-02-27 10:51:19.710660 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/4214309798 luksKillSlot /var/lib/ceph/osd/block-tmp 0
2023-02-27 10:51:23.094666 I | cephosd: adding new key to slot "0" of the device "/var/lib/ceph/osd/block-tmp"
2023-02-27 10:51:23.095229 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/104985237 --key-slot=0 luksAddKey /var/lib/ceph/osd/block-tmp /tmp/1275894378
2023-02-27 10:51:33.981403 I | cephosd: removing key slot "0", if found, of the device "/var/lib/ceph/osd/block.db-tmp"
2023-02-27 10:51:33.982012 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/2042354239 luksKillSlot /var/lib/ceph/osd/block.db-tmp 0
2023-02-27 10:51:36.526831 I | cephosd: adding new key to slot "0" of the device "/var/lib/ceph/osd/block.db-tmp"
2023-02-27 10:51:36.527194 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/444230199 --key-slot=0 luksAddKey /var/lib/ceph/osd/block.db-tmp /tmp/1825221398
2023-02-27 10:52:01.612414 I | cephosd: removing key slot "0", if found, of the device "/var/lib/ceph/osd/block.wal-tmp"
2023-02-27 10:52:01.617195 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/1125153552 luksKillSlot /var/lib/ceph/osd/block.wal-tmp 0
2023-02-27 10:52:04.531993 I | cephosd: adding new key to slot "0" of the device "/var/lib/ceph/osd/block.wal-tmp"
2023-02-27 10:52:04.532709 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/3174693213 --key-slot=0 luksAddKey /var/lib/ceph/osd/block.wal-tmp /tmp/2128256523
2023-02-27 10:52:22.727476 I | cephosd: updating the new key in the KMS
2023-02-27 10:52:22.756352 I | cephosd: fetching the key from the KMS to verify it.
2023-02-27 10:52:22.759071 I | cephosd: removing the old key from the slot "1" of the device "/var/lib/ceph/osd/block-tmp"
2023-02-27 10:52:22.759371 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/1029750050 luksKillSlot /var/lib/ceph/osd/block-tmp 1
2023-02-27 10:52:24.448413 I | cephosd: removing the old key from the slot "1" of the device "/var/lib/ceph/osd/block.db-tmp"
2023-02-27 10:52:24.448854 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/3894365656 luksKillSlot /var/lib/ceph/osd/block.db-tmp 1
2023-02-27 10:52:26.643961 I | cephosd: removing the old key from the slot "1" of the device "/var/lib/ceph/osd/block.wal-tmp"
2023-02-27 10:52:26.644270 D | exec: Running command: cryptsetup --verbose --key-file=/tmp/2091646819 luksKillSlot /var/lib/ceph/osd/block.wal-tmp 1
2023-02-27 10:52:29.956407 I | cephosd: Successfully rotated the key

```
</details>

Rakshith-R commented on cmd/rook/secret.go:
```diff
@@ -113,3 +121,33 @@ func getSecret(cmd *cobra.Command, args []string) {
+	cmd := &cobra.Command{
+		Use:   "rotate-key [kms-secret-key] [data-device] [metadata-device] [wal-device]",
+		Short: "Rotate key",
+		Args:  cobra.RangeArgs(2, 4),
```
Signifies that it takes minimum 2 and maximum 4 arguments.

https://pkg.go.dev/github.com/spf13/cobra#RangeArgs

Rakshith-R commented on deploy/charts/rook-ceph/templates/clusterrole.yaml:
```diff
@@ -207,6 +208,14 @@ rules:
+  # cron jobs.
+  - deployments/finalizers
+  verbs:
+  - update
```
> This doesn't require the `create` in addition to `update`? I thought we would set the owner ref during creation too.

This is for setting ownerref of cronjob onto the OSD deployment.
When OSD deployment is deleted, its key rotation cronjob also gets deleted.
Update is enough for this case.

Rakshith-R commented on deploy/examples/common-second-cluster.yaml:
```diff
@@ -166,6 +166,11 @@ metadata:
   name: rook-ceph-osd
   namespace: $NAMESPACE
 rules:
+  # this is needed for rook's "key-management" CLI to fetch the vault token from the secret when
```
That part was from the original wording in file `deploy/charts/library/templates/_cluster-role.tpl` above.

Even the required `secret- get` rbac was missing in common-second-cluster.yaml. 
This change fixes that too.

Rakshith-R commented on pkg/daemon/ceph/osd/key_rotation.go:
```diff
@@ -0,0 +1,94 @@
+	// Fetch key to verify its the new key.
+	keyInKMS, err := kms.GetSecret(secretName)
+	if keyInKMS != newKey {
+		return err
```
checked all the places and modified where ever necessary :+1:

Rakshith-R commented on .github/workflows/canary-integration-test.yml:
```diff
@@ -792,6 +787,7 @@ jobs:
           cat tests/manifests/test-on-pvc-db.yaml >> tests/manifests/test-cluster-on-pvc-encrypted.yaml
           cat tests/manifests/test-on-pvc-wal.yaml >> tests/manifests/test-cluster-on-pvc-encrypted.yaml
           kubectl create -f tests/manifests/test-cluster-on-pvc-encrypted.yaml
+          kubectl patch -n rook-ceph cephcluster rook-ceph --type merge -p '{"spec":{"security":{"keyRotation":{"enabled": true, "schedule":"*/5 * * * *"}}}}'
```
> How often will the key rotation be tested here? 5 times per minute?

once every 5 minutes. https://crontab.guru/#*/5_*_*_*_* 

the script function `verify_key_rotation` 
checks secret for change in key value every 20 seconds for 10 minutes. 
once it is changed, 
`test osd deployment removal and re-hydration` test will take care of check of opening encrypted device with new key.

Rakshith-R commented on pkg/daemon/ceph/osd/key_rotation.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	// Ensure currentKey is in slot 1.
+	for _, devicePath := range devicePaths {
```
> If we have one key rotation job per OSD, why would we ever have more than one device path? Shouldn't this method just take a single devicePath?

This refers to osd data, metadata and wal devices, 
when these are based on separate pvcs, each encrypted device needs to be updated with new keys.

Rakshith-R commented on pkg/operator/ceph/cluster/osd/key_rotation.go:
```diff
@@ -0,0 +1,284 @@
+	logger.Infof("successfully started OSD key rotation cron jobs")
+
+	return nil
+}
```
cronjob has support for this. 
By default, currently it keeps around 3 successful pod history and 1 failed pod around.
https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#jobs-history-limits

Rakshith-R commented on pkg/daemon/ceph/osd/kms/k8s.go:
```diff
@@ -54,6 +54,24 @@ func (c *Config) storeSecretInKubernetes(pvcName, key string) error {
+	secret.StringData = map[string]string{OsdEncryptionSecretNameKeyName: key}
+	// Update the Kubernetes Secret
+	_, err = c.context.Clientset.CoreV1().Secrets(c.ClusterInfo.Namespace).Update(c.ClusterInfo.Context, secret, metav1.UpdateOptions{})
+	if err != nil && !kerrors.IsAlreadyExists(err) {
```
yes, its not required. I must have missed removing it from my earlier trials, removed it now.
thanks.
